typed/symdiff_test.go

```diff
@@ -0,0 +1,383 @@
+        list:
+          elementType:
+            untyped: {}
+          elementRelationship: atomic
+`,
+	triplets: []mergeTriplet{{
+		`{}`,
+		`{"inner":[]}`,
+		`{"inner":[]}`,
+	}, {
+		`{}`,
+		`{"inner":null}`,
+		`{"inner":null}`,
+	}, {
+		`{"inner":null}`,
+		`{"inner":[]}`,
+		`{"inner":[]}`,
+	}, {
+		`{"inner":[]}`,
+		`{"inner":null}`,
+		`{"inner":null}`,
+	}, {
+		`{"inner":[]}`,
+		`{"inner":[]}`,
+		`null`,
+	}},
+}, {
+	name:         "struct grab bag",
+	rootTypeName: "myStruct",
+	schema: `types:
+- name: myStruct
+  struct:
+    fields:
+    - name: numeric
+      type:
+        scalar: numeric
+    - name: string
+      type:
+        scalar: string
+    - name: bool
+      type:
+        scalar: boolean
+    - name: setStr
+      type:
+        list:
+          elementType:
+            scalar: string
+          elementRelationship: associative
+    - name: setBool
+      type:
+        list:
+          elementType:
+            scalar: boolean
+          elementRelationship: associative
+    - name: setNumeric
+      type:
+        list:
+          elementType:
+            scalar: numeric
+          elementRelationship: associative
+`,
+	triplets: []mergeTriplet{{
+		`{"numeric":1}`,
+		`{"numeric":3.14159}`,
+		`{"numeric":3.14159}`,
+	}, {
+		`{"numeric":3.14159}`,
+		`{"numeric":1}`,
+		`{"numeric":1}`,
+	}, {
+		`{"string":"aoeu"}`,
+		`{"bool":true}`,
+		`{"string":"aoeu","bool":true}`,
+	}, {
+		`{"setStr":["a","b","c"]}`,
+		`{"setStr":["a","b"]}`,
+		`{"setStr":["c"]}`,
+	}, {
+		`{"setStr":["a","b"]}`,
+		`{"setStr":["a","b","c"]}`,
+		`{"setStr":["c"]}`,
+	}, {
+		`{"setStr":["a","b","c"]}`,
+		`{"setStr":[]}`,
+		`{"setStr":["a","b","c"]}`,
+	}, {
+		`{"setStr":[]}`,
+		`{"setStr":["a","b","c"]}`,
+		`{"setStr":["a","b","c"]}`,
+	}, {
+		`{"setBool":[true]}`,
+		`{"setBool":[false]}`,
+		`{"setBool":[true,false]}`,
+	}, {
+		`{"setNumeric":[1,2,3.14159]}`,
+		`{"setNumeric":[1,2,3]}`,
+		// KNOWN BUG: this order is wrong
+		`{"setNumeric":[3.14159,3]}`,
+	}},
+}, {
+	name:         "associative list",
+	rootTypeName: "myRoot",
+	schema: `types:
+- name: myRoot
+  struct:
+    fields:
+    - name: list
+      type:
+        namedType: myList
+    - name: atomicList
+      type:
+        namedType: mySequence
+- name: myList
+  list:
+    elementType:
+      namedType: myElement
+    elementRelationship: associative
+    keys:
+    - key
+    - id
+- name: mySequence
+  list:
+    elementType:
+      scalar: string
+    elementRelationship: atomic
+- name: myElement
+  struct:
+    fields:
+    - name: key
+      type:
+        scalar: string
+    - name: id
+      type:
+        scalar: numeric
+    - name: value
+      type:
+        namedType: myValue
+    - name: bv
+      type:
+        scalar: boolean
+    - name: nv
+      type:
+        scalar: numeric
+- name: myValue
+  map:
+    elementType:
+      scalar: string
+`,
+	triplets: []mergeTriplet{{
+		`{"list":[{"key":"a","id":1,"value":{"a":"a"}}]}`,
```

Comment:
Can you add a test that adds a value to an existing item? I suspect if we build a symmetric difference like this, it's going to be hard to identify which list items have been added as opposed to changed.

Reply:
Test added now